internal/beatcmd/beat.go

```diff
@@ -457,6 +491,156 @@ func (b *Beat) registerMetrics() {
+	case metricdata.Gauge[int64]:
+		if len(data.DataPoints) != 1 || data.DataPoints[0].Attributes.Len() != 0 {
+			break
+		}
+		return data.DataPoints[0].Value, true
+	}
+	return 0, false
+}
+
+// Adapt go-docappender's OTel metrics to beats stack monitoring metrics,
+// with a mixture of libbeat-specific and apm-server specific metric names.
+func addDocappenderLibbeatOutputMetrics(ctx context.Context, v monitoring.Visitor, sm metricdata.ScopeMetrics) {
+	for _, m := range sm.Metrics {
+		switch m.Name {
+		case "elasticsearch.events.processed":
+			var acked, toomany, failed int64
+			data, _ := m.Data.(metricdata.Sum[int64])
+			for _, dp := range data.DataPoints {
+				status, ok := dp.Attributes.Value(attribute.Key("status"))
+				if !ok {
+					continue
+				}
+				switch status.AsString() {
+				case "Success":
+					acked++
+				case "TooMany":
+					toomany++
+					fallthrough
+				default:
+					failed++
+				}
+			}
+			monitoring.ReportInt(v, "events.acked", acked)
+			monitoring.ReportInt(v, "events.failed", failed)
+			monitoring.ReportInt(v, "events.toomany", toomany)
+		case "elasticsearch.events.count":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "events.total", value)
+			}
+		case "elasticsearch.events.queued":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "events.active", value)
+			}
+		case "elasticsearch.flushed.bytes":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "write.bytes", value)
+			}
+		case "elasticsearch.bulk_requests.count":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "events.batches", value)
+			}
+		}
+	}
+}
+
+func addDocappenderLibbeatPipelineMetrics(ctx context.Context, v monitoring.Visitor, sm metricdata.ScopeMetrics) {
+	for _, m := range sm.Metrics {
+		switch m.Name {
+		case "elasticsearch.events.count":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "events.total", value)
+			}
+		}
+	}
+}
+
+// Add non-libbeat Elasticsearch output metrics under "output.elasticsearch".
+func addDocappenderOutputElasticsearchMetrics(ctx context.Context, v monitoring.Visitor, sm metricdata.ScopeMetrics) {
+	for _, m := range sm.Metrics {
+		switch m.Name {
+		case "elasticsearch.bulk_requests.count":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "bulk_requests.completed", value)
+			}
+		case "elasticsearch.bulk_requests.available":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "bulk_requests.available", value)
+			}
+		case "elasticsearch.indexers.created":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "indexers.created", value)
+			}
+		case "elasticsearch.indexers.destroyed":
+			if value, ok := getScalarInt64(m.Data); ok {
+				monitoring.ReportInt(v, "indexers.destroyed", value)
+			}
+			// TODO output.elasticsearch.indexers.active (created - destroyed?)
```

Comment:
Continuing from https://github.com/elastic/apm-server/pull/15094#discussion_r1927668560

> this is not produced by the go-docappender, it's one of the legacy metrics. If we think it's useful we should add it there. WDYT ?

It might be a good idea to add to go-docappender, but I'm not sure that we can wait for that. Can we calculate it here as created-destroyed, and add a TODO to either add it to go-docappender or remove the metric?

Reply:
I've added it back (should be created-destroyed+1), and we can think about whether we want to remove it or add it to go-docappender later.